### ISSUE TITLE
Roadmap 2024 (discussion)

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#58392]
This is ClickHouse roadmap 2024.
Descriptions and links are to be filled.

This roadmap does not cover the tasks related to infrastructure, orchestration, documentation, marketing, external integrations, drivers, etc.

See also:

Roadmap 2023: https://github.com/ClickHouse/ClickHouse/issues/44767
Roadmap 2022: https://github.com/ClickHouse/ClickHouse/issues/32513
Roadmap 2021: https://github.com/ClickHouse/ClickHouse/issues/17623
Roadmap 2020: [link](https://github.com/ClickHouse/ClickHouse/blob/be29057de1835f6f4a17e03a422b45b81efe6833/docs/ru/whats-new/extended-roadmap.md)

## SQL Compatibility

✔️ Enable Analyzer by default
Non-constant CASE, non-constant IN
Remove old predicate pushdown mechanics
Correlated subqueries with decorrelation
Transforming anti-join: LEFT JOIN ... WHERE ... IS NULL to NOT IN
Deriving index condition from the right-hand side of INNER JOIN
✔️ JOINs reordering and extended pushdown
Time data type

## Data Storage

Userspace page cache
✔️ Adaptive mode for asynchronous inserts
✔️ Semistructured Data: Variant data type
Semistructured Data: Sharded Maps
✔️ Semistructured Data: JSON data type
Transactions for Replicated tables
Lightweight Updates v2
Uniform treatment of LowCardinality, Sparse, and Const columns
✔️ Settings to control the consistency of projections on updates
Replicated Catalog :cloud:
✔️ On-disk storage for Keeper
Query cache on disk
✔️ Decoupling of object storages and metadata
Full-text indices (production readiness)
Vector search indices (production readiness)

## Security, access control, and isolation

✔️ Definers (encapsulation of access control) for views
✔️ Warnings and limits on the number of database objects
Dynamic configuration of query handlers
JWT authentication :cloud:
Data masking in row-level security :cloud:
Secure storage for named collections :cloud:
Cancellation points for long operations
✔️ Resource scheduler (continuation)

## Query Processing

Parallel replicas with task callbacks (production readiness)
Parallel replicas with parallel distributed INSERT SELECT
Automatic usage of -Cluster table functions
✔️ Adaptive thresholds for data spilling on disk
✔️ Optimization with subcolumns by default

## Interfaces & External Data

✔️ Support for Iceberg Data Catalog
✔️ Support for Hive-style partitioning
Explicit queries in external tables
Even simpler data upload
HTTP API for simple query construction
Unification of data lake and file-like functions

## Testing & Hardening

✔️ Revive coverage
Fuzzer of data formats
Fuzzer of network protocols
Server-side AST query fuzzer
✔️ Generic fuzzer for query text
Randomization of DETACH/ATTACH in tests
Integration with SQLSmith
Embedded documentation

## Experiments & Research

Cancelled: Multi-RAFT for Keeper
Cancelled: MaterializedPostgreSQL (production readiness)
SSH protocol for the server
Support for PromQL
Streaming queries
Freeform text format
Key-value data marts
Decouple of columns and buffers
Lazy reading of ranges
Instant attaching tables from backups
An object storage to borrow space from the filesystem cache
COW disks
ALTER PRIMARY KEY
Autocompletion with language models
Decentralized tables
Unique Key Constraint

---
The roadmap covers the top focus items for both external contributors and full-time ClickHouse employees.
The items marked with the :cloud: icon are meant for ClickHouse Cloud (proprietary).
We expect 50..80% completion of the roadmap according to the results from previous years.

**Comment by alanpaulkwan** [comment]
"HTTP API for simple query construction" - would be really awesome if Python/R/DuckDB could read an arbitrary output like S3 or just a file download. Amazing.

**Comment by chenziliang** [comment]
Would like to see streaming processing become main stream in ClickHouse :)

**Comment by ahmed-adly-khalil** [comment]
Would be great to see deeper NATS integration, mainly using JWT auth.

**Comment by ucasfl** [comment]
> Support for Iceberg Data Catalog

`Catalog` is a good concept, but if want to introduce `catalog` into ClickHouse, we need refactor current metadata's structure, from `Database -> Table` to `Catalog -> Database -> Table`, then all tables created with builtin engine will under a `internal_catalog`, and we can have `iceberg_catalog`, `hive_catalog`. Not sure it's a good idea.

Besides, we still face the difficulty that we don't have Iceberg C++ API.

**Comment by olly-writes-code** [comment]
Hi! Adding a request for more GIS features. One important one is the ability to transform a geometry to a specified SRID. Redshift docs [here](https://docs.aws.amazon.com/redshift/latest/dg/ST_Transform-function.html)

**Comment by alexey-milovidov** [comment]
@ucasfl, we can support mapping a specific database from the Iceberg data catalog as a database in ClickHouse.
In this way, we don't have to map a whole catalog at once but allow doing it database by database.

**Comment by alexey-milovidov** [comment]
@ahmed-adly-khalil, while NATS is not on the main list (but nice to have), there are some items that we already started to do: https://github.com/ClickHouse/ClickHouse/issues/39459

**Comment by alexey-milovidov** [comment]
@alanpaulkwan, yes, in the simplest form, it represents a table like a file: https://github.com/ClickHouse/ClickHouse/issues/46925 but also allows to customize the result.

**Comment by ucasfl** [comment]
> @ucasfl, we can support mapping a specific database from the Iceberg data catalog as a database in ClickHouse. In this way, we don't have to map a whole catalog at once but allow doing it database by database.

Yes, external database like `DataMySQL` also a good idea.

In fact, we already implemented external database named `DatabaseIceberg` and used in our real user cases in my company, but due to the lack of Iceberg C++ API, in our implementation, we introduce a external service to help finished things like partition prune, min-max index prune, also provide function of metadata cache, manifest cache. The main infra is like below:
![image](https://github.com/ClickHouse/ClickHouse/assets/22127746/a58ccc69-1851-49b4-aa03-982b7785ea5a)

![image](https://github.com/ClickHouse/ClickHouse/assets/22127746/69aaa146-2e0b-4aa4-a1ff-9aeb0e1bfb99)

Without external service and  Iceberg C++ API, I think it's not easy to implement effective Iceberg table reading.

**Comment by jrdi** [comment]
Would be great to know if there are plans to keep working on improving zero-copy and Cloud Storage during 2024. I'm constantly seeing improvements and bug fixes. Will we see zero-copy ready for production this year?

> Decoupling of object storages and metadata

Does this mean moving metadata from disks to Keeper or any other shared store?

**Comment by alexey-milovidov** [comment]
@jrdi 

> Would be great to know if there are plans to keep working on improving zero-copy and Cloud Storage during 2024. I'm constantly seeing improvements and bug fixes. Will we see zero-copy ready for production this year?

We have to fix the issues in zero-copy replication because it is still tested in CI, and used in production on older services in ClickHouse Cloud. For example, issues like this are found: https://github.com/ClickHouse/ClickHouse/pull/58333. But the track record of zero-copy replication is not good, and we expect to stop using it, then remove it from CI, and keep it on life support without further changes.

> Does this mean moving metadata from disks to Keeper or any other shared store?

This is https://github.com/ClickHouse/ClickHouse/pull/58357

We currently have the following metadata options:
1. Metadata on local filesystem.
2. No separate metadata.
3. Metadata in .index files in directories.
4. Metadata in a backup.
5. Metadata in Keeper.
6. and more to come, e.g. https://github.com/ClickHouse/ClickHouse/issues/58347

And, we have the following object storage options:
1. S3.
2. HDFS.
3. Azure.
4. Web.
5. Local filesystem.
6. Borrowing space from the filesystem cache.

The task is to allow the cross-product of these options.

**Comment by chenziliang** [comment]
> Would be great to see deeper NATS integration, mainly using JWT auth.

@ahmed-adly-khalil may I ask if you like streaming processing / analytics against NATS via ClickHouse ?

**Comment by jrdi** [comment]
Thanks, @alexey-milovidov!

> But the track record of zero-copy replication is not good, and we expect to stop using it, then remove it from CI, and keep it on life support without further changes.

I can understand this decision but it's a pity. This mean that open source version won't have a productive method to separate compute and storage. Do you think this could change in the short/mid term? Even something like a plan with CH help and guidance on improvements that can be done to keep support by external contributors sounds better than keeping the feature out of the CI.

**Comment by alexey-milovidov** [comment]
It is not guaranteed and not in the plans, but we might have an implementation in the future- the only thing for sure is that it will not be based on zero-copy replication.

**Comment by bputt-e** [comment]
`Unique Key Constraint` would be great, could remove our deduplication step in our processing pipeline

**Comment by mbtolou** [comment]
`Unique Key Constraint` is great idea.

**Comment by alanpaulkwan** [comment]
I really lke the unique key idea - hope it can folow ReplacingMergeTree and allow user to decide which to keep. For me the options seem to be (1) incumbent data entry, (2) newest data entry, (3) an integer describing version priority. I've created arbitrary values to keep the "best value".

**Comment by mwarkentin** [comment]
@jrdi theres a proposal from Altinity here: https://github.com/ClickHouse/ClickHouse/issues/54644

**Comment by earlev4** [comment]
**_A huge thank you to ClickHouse team and all the contributors for the amazing work on ClickHouse! I sincerely appreciate it._** 

Just curious. The [Roadmap 2023](https://github.com/ClickHouse/ClickHouse/issues/44767) mentioned a "Recursive CTE" task, but I do not see it mentioned in the Roadmap 2024. Are plans to implement recursive CTEs in the future?

Thanks again!

**Comment by alexey-milovidov** [comment]
@earlev4, It was planned for the previous year after enabling Analyzer, but we didn't manage to enable Analyzer under that schedule, so I've added it as the major item for 2024, but I'm afraid to add recursive CTEs. We are considering it for implementation, but not on the list of main items.

**Comment by earlev4** [comment]
Thanks so much, @alexey-milovidov! I sincerely appreciate the detailed response. It is very helpful. I am very grateful to you and the team for ClickHouse!

**Comment by domainio** [comment]
Hi, is there a plan to support Apache Iceberg writing with MERGE operation?

**Comment by zheyu001** [comment]
Is there any chance to support iceberg v2?

**Comment by 1392657590** [comment]
Do you have time to resolve  In high-concurrency scenarios, the performance of ClickHouse Keeper is lower than that of ZooKeeper.We found this issue when replacing zk with keeper. The replacement plan has been temporarily suspended

**Comment by JackyWoo** [comment]
@1392657590 maybe you can try RaftKeeper

**Comment by jiugem** [comment]
I don't see MaterializedMySQL in the Roadmap

**Comment by zhanglistar** [comment]
@alexey-milovidov What about non-equal join, and asof join with muliple conditions? Any plan? Thanks.

**Comment by chrisgoddard** [comment]
Would be super excited to see production support for Vector search indices - especially on Clickhouse Cloud. Every week there seems like there's a different vector database and I can't wait until I can just use Clickhouse for *everything*

**Comment by xevix** [comment]
I would have major performance bottlenecks alleviated with materialized CTEs, to avoid DB roundtrips to create many intermediate results in tables. DuckDB recently added this last year which was great to see, was wondering if Clickhouse was thinking about this as well: https://github.com/ClickHouse/ClickHouse/issues/53449.

Fantastic work on recent releases, with usability features like `ORDER BY ALL` making it faster to query things ad hoc, and tight S3 integration with system credentials just magically pulled in 🎉

**Comment by guoxiaolongzte** [comment]
Support for Hive style partitioning.

Does it support dynamic hive partition writing?

**Comment by guoxiaolongzte** [comment]
> > Support for Iceberg Data Catalog
> 
> `Catalog` is a good concept, but if want to introduce `catalog` into ClickHouse, we need refactor current metadata's structure, from `Database -> Table` to `Catalog -> Database -> Table`, then all tables created with builtin engine will under a `internal_catalog`, and we can have `iceberg_catalog`, `hive_catalog`. Not sure it's a good idea.
> 
> Besides, we still face the difficulty that we don't have Iceberg C++ API.

@alexey-milovidov
When do we plan to support the hive catalog and hudi catalog?

**Comment by mingmwang** [comment]
> @1392657590 maybe you can try RaftKeeper

@JackyWoo 

Do you have some benchmark data to share between the performance and throughput of ClickHouse Keeper and RaftKeeper ?

**Comment by JackyWoo** [comment]
@mingmwang we did not compare them right  now.

**Comment by JackyWoo** [comment]
@mingmwang we did not compare them right now.

**Comment by softiger** [comment]
@JackyWoo could you share the comparison result between Zookeeper and RaftKeeper? I'm really interested in it, thanks!

**Comment by JackyWoo** [comment]
@softiger You can find it [here](https://github.com/JDRaftKeeper/RaftKeeper/blob/master/benchmark%2FBenchmark.md).

**Comment by immelnikoff** [comment]
I would like to see binary search for pre-ordered arrays in the next functions:
has()
hasAny()
hasAll()
arrayIntersect().
In general, I would like to see accelerated functions for pre-ordered arrays.

**Comment by wordhardqi** [comment]
CK could plan query optimizer for complex queries.

**Comment by wordhardqi** [comment]
Starrocks, Snowflake and Byconity have it.

**Comment by Dileep-Dora** [comment]
I see from [2023 roadmap](https://github.com/ClickHouse/ClickHouse/issues/44767#issuecomment-1764921315) inverted indices implementation is not a priority. https://github.com/ClickHouse/ClickHouse/pull/38667 .

are we considering this for this year or any other plans on improving text search performance.

**Comment by alexey-milovidov** [comment]
So far, there is a prototype implementation of inverted indices (you can unlock it with `allow_experimental_inverted_index`) - it is not ready and should not be used in production. It was not tested on realistic datasets.

**Comment by Dileep-Dora** [comment]
@alexey-milovidov thanks for the reply, yes we've tried this experimental feature but performance was not upto the mark. hence checking do we have any plans for this in 2024

**Comment by johnpyp** [comment]
Curious about prioritizing supporting `ORDER BY` optimizations for projected views? This is the one thing holding my team back from using Clickhouse for `WHERE` query usecases where we want to replicate the ease-of-use and flexibility of traditional database indices.

We'd love to be able to create potentially many projected views on top of one table with different combinations of different ORDER BY query optimizations *and* WHERE query optimizations.

**Comment by anvaari** [comment]
Is there any plan to support deserializing Protobuf through a schema registry? It's needed.

**Comment by callicles** [comment]
On Full-test search indices - will that include things like TF-IDF, lemmatization, stop word removal etc... (mentioned in the blog post)

**Comment by ahmed-adly-khalil** [comment]
can we please keep the JSON type https://clickhouse.com/docs/en/sql-reference/data-types/json ? it's been useful to us in many ways and it's sad it's going to be removed, ref: https://clickhousedb.slack.com/archives/CU478UEQZ/p1717653663312899

**Comment by ahmed-adly-khalil** [comment]
> can we please keep the JSON type https://clickhouse.com/docs/en/sql-reference/data-types/json ? it's been useful to us in many ways and it's sad it's going to be removed, ref: https://clickhousedb.slack.com/archives/CU478UEQZ/p1717653663312899

correcting myself: there is amazing work is being done to revamp JSON type, details here: https://github.com/ClickHouse/ClickHouse/issues/54864

**Comment by alexey-milovidov** [comment]
@earlev4 Recursive CTEs have been added in 24.4.

**Comment by dragon-wzl** [comment]
@alexey-milovidov  Is Lightweight Updates v2 a feature of ClickHouse Cloud or the community version?

**Comment by alexey-milovidov** [comment]
Lightweight updates are both for the Cloud and open-source.

**Comment by jiebinn** [comment]
Do we have a plan to support `bfloat16` data type in schema for vector search?

**Comment by alexey-milovidov** [comment]
@jiebinn, yes, this task is in work.

**Comment by alberttwong** [comment]
> > > Support for Iceberg Data Catalog
> > 
> > 
> > `Catalog` is a good concept, but if want to introduce `catalog` into ClickHouse, we need refactor current metadata's structure, from `Database -> Table` to `Catalog -> Database -> Table`, then all tables created with builtin engine will under a `internal_catalog`, and we can have `iceberg_catalog`, `hive_catalog`. Not sure it's a good idea.
> > Besides, we still face the difficulty that we don't have Iceberg C++ API.
> 
> @alexey-milovidov When do we plan to support the hive catalog and hudi catalog?

Apache Hudi is working on a rust implementation.

**Comment by franz101** [comment]
Simpler deduplication for replacingmergetree. LIMIT 1 BY id works but causes some headache with sorting.

**Comment by franz101** [comment]
Cascading deletes

**Comment by lichehuo** [comment]
which version will support the feature of full text index

**Comment by coffenbacher** [comment]
Hi team - what does the [PeerDB acquisition](https://clickhouse.com/blog/clickhouse-welcomes-peerdb-adding-the-fastest-postgres-cdc-to-the-fastest-olap-database?loc=eyebrow) mean for the future of MaterializedPostgreSQL? We have loved MaterializedPostgreSQL but we run into continuous bugs with it (all noted by others in the github issue tracker for some time). I am trying to decide what direction to go from here.

**Comment by alexey-milovidov** [comment]
@lichehuo, the full-text index is not under development, and the existing experimental implementation is nowhere close to production. Most likely, it will have to be postponed.

@coffenbacher, MaterializedPostgreSQL continues to remain in the experimental state - so far, there are not many efforts to improve it, but it will continue to exist.

**Comment by saisrirampur** [comment]
Thanks for reaching out, @coffenbacher, and nice to meet you!

From a PeerDB perspective, it is production-ready and heavily optimized for Postgres/ClickHouse. We are working on integrating it into ClickPipes (the native ingestion service for ClickHouse Cloud) for native Postgres CDC. In the meantime, you can either use the [OSS](https://github.com/PeerDB-io/peerdb) or [Cloud](https://auth.peerdb.cloud/en/signup) (comes with a free trial) options. The Cloud option would be more suitable if you are a ClickHouse Cloud user. :)

**Comment by the4thamigo-uk** [comment]
> Lightweight updates are both for the Cloud and open-source.

@alexey-milovidov do you have a rough idea when this will become available in open-source?

**Comment by alexey-milovidov** [comment]
There are two components of the lightweight updates feature.
The first is named "on-the-fly mutations," and the second is "patch parts."
Currently, we are preparing to release on-the-fly mutations in open-source.
It is likely to happen in the nearest months.

**Comment by leutrimmusliuu** [comment]
@alexey-milovidov, it would be incredibly valuable to have a MergeTree engine in ClickHouse similar to ReplacingMergeTree, but with guaranteed row replacement based on user-specified unique keys during inserts. This feature, similar to the functionality of EmbeddedRocksDb but with better query performance, is something ClickHouse currently lacks. It would be especially beneficial for managing dimension tables, where maintaining unique rows is critical. Also this would allow us just insert on the dimension table, without having to deal with deduplication (without maintaining signs or versions) .

**Comment by den-crane** [comment]
@leutrimmusliuu you should use FINAL 
https://clickhouse.com/docs/en/sql-reference/statements/select/from#final-modifier
https://clickhouse.com/docs/en/operations/settings/settings#final

>guaranteed row replacement based on user-specified unique keys during inserts.

* https://github.com/ClickHouse/ClickHouse/pull/44534

**Comment by leutrimmusliuu** [comment]
@den-crane  It's generally not recommended to use FINAL on larger tables due to the potential for merges occurring during query execution. The UniqueMergeTree pull request seems to address this issue, and I hope it gets completed in the future.

**Comment by den-crane** [comment]
@leutrimmusliuu, 

>especially beneficial for managing dimension tables

Is not general case and not common for big-data. 
It's very specific. I use FINAL for dimension tables (up to 50mil. rows). Works like a charm.

**Comment by Nyakutka** [comment]
Will it be possible to create nullable columt with composite type e.g. `Nullable(Array(string))`?
Does this limitation depend on some bug?

**Comment by mkarbo** [comment]
Any updates on data masking policies?

**Comment by the4thamigo-uk** [comment]
I'm wondering if there is an better estimate for when lightweight update will become available on open source?

**Comment by abhimanyusinghgaur** [comment]
Would love to see a parallel version of `WITH FILL` implemented with a sorting prefix and interpolation. Currently, the below part in a query executes sequentially:
```
order by c1, c2, c3, date ASC WITH FILL FROM toDate('start-of-month') TO toDate('end-of-month') STEP toIntervalDay(1) INTERPOLATE (c5)
```
That ends up consuming a lot of time. My dataset has ~340 million rows at the `start-of-month` and the result after filling has around ~10.5 billion rows. It took ~30mins to run this fill on a machine with `AMD EPYC 7502P 32-Core Processor (64 threads, 256gb RAM)` simply because it was only using 1 CPU. If the `WITH FILL` can be executed in parallel, that would be a really nice thing to have.
Not sure if #33203  is somehow related, but I think filling can be made parallel for each partition. In my case, the order by key is the partition.

**Comment by gitlijian** [comment]
@mwarkentin @jrdi @coffenbacher @xevix @alberttwong Hi. Are there any tasks that haven't been completed yet? Can you give me some?

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/issues/74046
